gh-128641: Fix ConfigParser.read Perfomance Regression

2trvl opened the pull request:
<!-- gh-issue-number: gh-128641 -->
* Issue: gh-128641
<!-- /gh-issue-number -->

2trvl committed:
perf: replace _Line with _LineParser

2trvl committed:
docs: add news entry

eendebakpt commented on Lib/configparser.py:
```diff
@@ -561,36 +561,32 @@ def __init__(self):
+            match = self.comments.inline.search(string)
+        if match:
+            return string[:match.start()].rstrip()
+        return string
```
```suggestion
        if self.comments.inline:
            match = self.comments.inline.search(string)
            if match:
                return string[:match.start()].rstrip()
        return string
```

eendebakpt commented on Misc/NEWS.d/next/Library/2025-02-02-23-47-35.gh-issue-128641.GFs673.rst:
```diff
@@ -0,0 +1,2 @@
+Reduce :mod:`configparser`. :class:`RawConfigParser`. :meth:`read` performance
```
These news entries are read by users that do not know the history. I would rewrite to something like `Improve performance of .... by up to ...%`.

Also the CI is failing on the links. Try something like:
```suggestion
Reduce :meth:`configparser.RawConfigParser.read` performance
```
(see https://devguide.python.org/documentation/markup/#roles)

2trvl committed:
refactor: apply requested changes

2trvl committed:
Merge branch 'python:main' into configparser_perf

2trvl committed:
docs: use ConfigParser instead of RawConfigParser

2trvl committed:
Merge branch 'configparser_perf' of https://github.com/2trvl/cpython into configparser_perf

jaraco commented on Lib/configparser.py:
```diff
@@ -659,9 +654,17 @@ def __init__(self, defaults=None, dict_type=_default_dict,
+            inline_comment_cre = None
+        self._comments = types.SimpleNamespace(
+            full=comment_prefixes,
+            inline=inline_comment_cre,
```
Previously, this code was just normalizing and saving some state in one expression. Now it's performing decision logic in three separate blocks. It's complicating the type, by making it asymmetric (full and inline now store different things) and non-deterministic (inline is now optional). Additionally, the logic for where this regular expression is used is now far away from where it's constructed; previously, the regular expression construction was near where it was used and documented by that method.

2trvl replied:
The number of blocks used for assignment has been reduced. The optionality of `inline_comment_cre` improves performance by not having to search for each line in an empty regex. Added comment about using cre in _LineParser.

jaraco commented on Lib/configparser.py:
```diff
@@ -561,36 +561,31 @@ def __init__(self):
+        return self._value
 
-    def _strip_inline(self):
-        """
-        Search for the earliest prefix at the beginning of the line or following a space.
-        """
```
This comment seems to have been lost.

2trvl replied:
Moved to `inline_comment_cre` assignment since it explains the regular expression.

jaraco commented on Lib/configparser.py:
```diff
@@ -561,36 +561,31 @@ def __init__(self):
+            match = self.comments.inline.search(string)
+            if match:
+                return string[:match.start()].rstrip()
+        return string
```
My preference instead of writing branching code is to write code that acts unconditionally, such as the previous regular expression, instead of tracking state (self.comments.inline is None) and altering the behavior based on state. I do this because it simplifies the analysis of the logic for correctness (fewer branches and states to track).
I can see how my preferred approach has potential performance disadvantages, such as when invoking the degenerate expression on most inputs.

2trvl replied:
Reduced cognitive complexity of this function. I don't use walrus operator because of your backport repository.

jaraco replied:
FTR, walrus operator was introduced in Python 3.8, so is suitable for backport.

In the latest revision, I've refactored the comment handling in _CommentSpec to encapsulate it and unify the approach, with no `if` statements needed.

jaraco commented on Lib/configparser.py:
```diff
@@ -1082,7 +1084,7 @@ def _read_inner(self, fp, fpname):
                 continue
 
-            first_nonspace = self.NONSPACECRE.search(line)
+            first_nonspace = self.NONSPACECRE.search(line.value)
```
I don't love that the value of the line can't be accessed directly, but must go through an attribute access. The previous string subclass gave much nicer ergonomics for a "line of text".

jaraco commented on Lib/configparser.py:
```diff
@@ -561,36 +561,31 @@ def __init__(self):
 
 
-class _Line(str):
+class _LineParser:
```
I saw some analysis in the ticket potentially implicating the use of an immutable object with a `__dict__`, so this change may be necessary, but I do think the previous implementation had some nicer ergonomics (i.e. a simple line of text with some specialized properties for different interpretations about that line). I'd like to explore options that retain those ergonomics.

2trvl replied:
Changing `str` to `object` made the code faster. GIL interpreter is slowed down by the creation of immutable objects. Free-threading also suffers from cachedproperty.

For loop in `_read_inner` became simpler by removing calling partial every iteration.

jaraco replied:
In 218db85a20877dd74aa9b8141fc8c313cc3a6d17, I've brought back the str subclass. It keeps the implementation easier to read and allows the `_Line` instance to represent the original text.

As you've observed, there is some performance effect. Here's the difference I see:

```
 cpython configparser_perf 🐚 docker run -it @$(docker build -q .) py -3.13 -m timeit -s 'import nixconfig' 'nixconfig.main()'
100 loops, best of 5: 2.33 msec per loop
 cpython configparser_perf 🐚 git checkout HEAD~1
HEAD is now at a098e697467 Rename variable from string to text. Use 'trimmed' to avoid masking the variable name.
 cpython a098e697467 🐚 docker run -it @$(docker build -q .) py -3.13 -m timeit -s 'import nixconfig' 'nixconfig.main()'
100 loops, best of 5: 2.13 msec per loop
```

There's about a 10% overhead to construct the `str` subclass. In my opinion, that's a reasonable tradeoff to avoid the indirection of `.value`. Still, maybe it's possible to avoid this str subclass. I'll keep exploring.

jaraco replied:
I tried a different approach, dropping the `_Line` class and storing the 'clean' value in a cache on the _Comment:

```diff
diff --git a/Lib/configparser.py b/Lib/configparser.py
index 4815d5399ba..9b5deba8f39 100644
--- a/Lib/configparser.py
+++ b/Lib/configparser.py
@@ -560,18 +560,6 @@ def __init__(self):
         self.errors = list()
 
 
-class _Line(str):
-    __slots__ = 'clean', 'has_comments'
-
-    def __new__(cls, val, *args, **kwargs):
-        return super().__new__(cls, val)
-
-    def __init__(self, val, comments):
-        trimmed = val.strip()
-        self.clean = comments.strip(trimmed)
-        self.has_comments = trimmed != self.clean
-
-
 class _CommentSpec:
     def __init__(self, full_prefixes, inline_prefixes):
         full_patterns = (
@@ -586,11 +574,12 @@ def __init__(self, full_prefixes, inline_prefixes):
             )
         self.pattern = re.compile('|'.join(itertools.chain(full_patterns, inline_patterns)))
 
+    @functools.cache
     def strip(self, text):
-        return self.pattern.sub('', text).rstrip()
+        return self.pattern.sub('', text.strip()).rstrip()
 
-    def load(self, text):
-        return _Line(text, self)
+    def in_(self, text):
+        return self.strip(text) != text.strip()
 
 
 class RawConfigParser(MutableMapping):
@@ -1062,12 +1051,12 @@ def _read(self, fp, fpname):
     def _read_inner(self, fp, fpname):
         st = _ReadState()
 
-        for st.lineno, line in enumerate(map(self._comments.load, fp), start=1):
-            if not line.clean:
+        for st.lineno, line in enumerate(fp, start=1):
+            if not self._comments.strip(line):
                 if self._empty_lines_in_values:
                     # add empty line to the value, but only if there was no
                     # comment on the line
-                    if (not line.has_comments and
+                    if (not self._comments.in_(line) and
                         st.cursect is not None and
                         st.optname and
                         st.cursect[st.optname] is not None):
@@ -1094,7 +1083,7 @@ def _handle_continuation_line(self, st, line, fpname):
         if is_continue:
             if st.cursect[st.optname] is None:
                 raise MultilineContinuationError(fpname, st.lineno, line)
-            st.cursect[st.optname].append(line.clean)
+            st.cursect[st.optname].append(self._comments.strip(line))
         return is_continue
 
     def _handle_rest(self, st, line, fpname):
@@ -1104,7 +1093,7 @@ def _handle_rest(self, st, line, fpname):
 
         st.indent_level = st.cur_indent_level
         # is it a section header?
-        mo = self.SECTCRE.match(line.clean)
+        mo = self.SECTCRE.match(self._comments.strip(line))
 
         if not mo and st.cursect is None:
             raise MissingSectionHeaderError(fpname, st.lineno, line)
@@ -1133,7 +1122,7 @@ def _handle_option(self, st, line, fpname):
         # an option line?
         st.indent_level = st.cur_indent_level
 
-        mo = self._optcre.match(line.clean)
+        mo = self._optcre.match(self._comments.strip(line))
         if not mo:
             # a non-fatal parsing error occurred. set up the
             # exception but keep going. the exception will be
```

But that resulted in worse performance (by about another 10%).

```
 cpython configparser_perf 🐚 docker run -it @$(docker build -q .) py -3.13 -m timeit -s 'import nixconfig' 'nixconfig.main()'
100 loops, best of 5: 2.52 msec per loop
```

jaraco reviewed (commented):
Thanks for this. The changes look sound and certainly help elucidate the reported performance degradation.

This biggest problem with this change is that it doesn't _capture_ the performance-sensitive concerns, so a future contributor (including myself having forgotten about these details) might refactor the code back to what it was (for example to reduce the number of logical branches or attribute accesses).

When I've encountered performance-sensitive parts of the code, I've done my best to (a) encapsulate the performance-sensitive aspects so they're identifiable and (b) include comments or docstrings to articulate their purpose. For example, in zipfile.Path, I added FastLookup:

https://github.com/python/cpython/blob/c53730171f18b90202aa0918b3c05412222bb1ec/Lib/zipfile/_path/__init__.py#L177-L193

This class encapsulated the performance optimizations, but also allows the concerns from the non-optimized version to be exposed separately (CompleteDirs). By disentangling these two concerns, it becomes more obvious why they're there and implemented that way. I'm not necessarily suggesting this pattern is appropriate here, though it could be.

The other thing I've done in these projects where possible is to include a regression test or benchmark that captures the performance expectations ([example](https://github.com/python/cpython/blob/8f9c6fa0773cff586d127d1b77b158edd4718035/Lib/test/test_zipfile/_path/test_complexity.py#L19-L107)). Unfortunately, often these sorts of tests require sophistication that isn't available in the stdlib. Sometimes, however, it's possible to put in a test that will capture serious performance issues, such as hanging on an O(n^2) operation.

I've left some comments to specific lines of the code indicating areas where I'm uneasy with the change (or might have liked to have done things differently). Because of this uneasiness, I'd like to explore other options. I'm not expecting you to simply address the comments.

I'm happy to help explore other options, but currently, it seems we don't have a good reproducer, so I'll ask for that in the ticket.

2trvl committed:
refactor: apply requested changes

2trvl committed:
Merge branch 'main' into configparser_perf

jaraco committed:
Move comment handling into a _CommentSpec class.

Rely on re.sub to perform the substitutions in a unified way across full and inline prefixes.

jaraco committed:
Remove the optimization, as it has no effect.

jaraco committed:
Rename variable from string to text. Use 'trimmed' to avoid masking the variable name.

jaraco committed:
Restored Line as str subclass.

Use 'slots' to avoid '__dict__'.

picnixz commented on Lib/configparser.py:
```diff
@@ -562,35 +561,36 @@ def __init__(self):
 class _Line(str):
+    __slots__ = 'clean', 'has_comments'
 
     def __new__(cls, val, *args, **kwargs):
```
Out of curiosity, why do we need an explicit `__new__`?

jaraco replied:
Without it, the superclass gets called with `str.__new__(val, encoding=comments)`.

picnixz replied:
Oh you're right.

picnixz commented on Lib/configparser.py:
```diff
@@ -562,35 +561,36 @@ def __init__(self):
+            # prefix at the beginning of the line or following a space
+            fr'(^|\s)({re.escape(prefix)}.*)'
+            for prefix in inline_prefixes
+            )
```
```suggestion
        )
```

Probably an oversight?

jaraco committed:
Normalize whitespace in diff.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

jaraco committed:
Merge branch 'main' into configparser_perf

jaraco reviewed (approved):
With these adjustments, I'm much happier with the approach and it regains the majority of the performance regression with a smaller diff and still mostly functional (paradigm) logic.

2trvl committed:
refactor: comments wrap
